### DISCUSSION TITLE
Turbo autocomplete search form equivalent for unpoly

### DISCUSSION THREAD

**Discussion by roscrl** [unpoly/unpoly#378] [Q&A] [answered]
Hi I'm wondering what the equivalent for this type of setup would be in Unpoly over Turbo.

It's a simple form which autocompletes as the user types. In the Turbo version I replace the inside of #search_results with the response from the server.

```
<form action="/stock/search" method="get" data-controller="stock-search"
      class="col-span-12 lg:col-start-5 lg:col-end-11 xl:col-start-6 xl:col-end-10 2xl:col-start-7 2xl:col-end-9 2xl:w-[24rem] 2xl:place-self-center flex flex-col relative">
    <input type="text" name="search" id="stock_search_input"
           data-action="stock-search#submit"
           data-stock-search-target="search"
           class="border rounded-md px-2 md:px-3 h-12 text-2xl lg:w-full lg:mx-auto"
           placeholder="APPL Apple..." autofocus="${autofocus}" autocomplete="off">
    <div class="flex justify-between lg:w-[24rem] lg:mx-auto mt-2 gap-1 sm:gap-1 pb-[0.7rem] overflow-x-auto "
         id="symbol_hotbar">
        <a href="/stock/AAPL"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-black focus-within:bg-black betterhover:hover:text-white">AAPL</a>
        <a href="/stock/MSFT"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-[#005DA6] betterhover:hover:text-white">MSFT</a>
        <a href="/stock/AMZN"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-[#FF9900] betterhover:hover:text-white">AMZN</a>
        <a href="/stock/TSLA"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-[#E31937] betterhover:hover:text-white">TSLA</a>
        <a href="/stock/BRK.B"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-[#2D2D94] betterhover:hover:text-white">BRK.B</a>
        <a href="/stock/JPM"
           class="border rounded-md px-2 sm:w-[4rem] text-center betterhover:hover:bg-[#503629] betterhover:hover:text-white">JPM</a>
    </div>

    <div id="search_results"
         class="absolute top-0 left-0 bg-white w-full translate-y-12 mt-1 font-semibold fast-fade z-10">
    </div>

</form>
```

Server side turbo-stream [response](https://github.com/roscrl/econicles/blob/main/src/main/resources/views/frames/stock_search.jte)

```
<turbo-stream action="update" target="search_results">
    <template>
        <div>
            <ol class="flex flex-col items-center gap-1">
                @for(Ticker ticker : tickers)
                    <a href="/stock/${ticker.symbol()}"
                       tabindex="0"
                       class="border w-full rounded-md flex justify-between items-center h-12 p-2 hover:bg-blue-50">
                        <span class="inline-block">${ticker.symbol()}</span>
                        <span class="ml-4 block truncate">${ticker.name()}</span>
                    </a>
                @endfor
            </ol>
        </div>
    </template>
</turbo-stream>
``` 

Also the [stimulus js controller](https://github.com/roscrl/econicles/blob/main/src/main/resources/views/controllers/stock_search_controller.js) manages some of the lifecycle.

I tried using ```up.submit(this.element, {target: '#search_results'});``` to stimulus js controller instead of requestSubmit but before even getting to the network response I see

![image](https://user-images.githubusercontent.com/13072760/167700670-de221518-c3db-4459-b4fe-d0d0117250bb.png)

**Comment by adam12** [comment, marked as answer]
To fix your error, try adding an id (or a surrogate `up-id`) to your `<form>` element. I'm guessing Unpoly doesn't like the square brackets in the `form[class]` attribute.

Other than that, your solution looks fine, and the Unpoly solution would be fairly similar to the Turbo solution. Inform Unpoly to submit the form, and then from the response, extract an element and replace the one in the DOM with the new element.

**Reply by roscrl** [reply]
Oh wow that was exactly the issue. I wonder why the tailwind css brackets interfere. Thank you!

**Reply by adam12** [reply]
We ran into something similar with the `:` in Tailwind a while back (see #169). I guess we'll need to update it to escape square brackets too. Glad to hear it worked.

**Comment by triskweline** [comment]
Relevant spec for CSS class names probably here: https://www.w3.org/TR/CSS2/syndata.html#characters
